File: src/cs/Naxp/Ast.cs
// Copyright (c) Tim Gordon.
// This file is licensed to you under the Apache Licence, Version 2.0. See the LICENSE file.

using System;
using System.Collections.Generic;

namespace LogMu;

/// <summary>
/// A node in the abstract syntax tree of a naxp.
/// </summary>
/// <remarks>
/// <para>
/// The tree keeps the structure the pattern was written in. Intervals and decimal ranges are
/// deliberately <em>not</em> expanded here. The cap on an interval count exists so
/// that an implementation can find a naxp invalid before expanding it, and expanding at parse time
/// would throw that away: <c>(A{99}){99}</c> is eleven characters of pattern and nearly ten
/// thousand characters of expansion.
/// </para>
/// <para>
/// Groups do not survive parsing. <c>(A)</c> and <c>A</c> give the same tree, and the two
/// abbreviated unified forms are expanded into the general one, since the specification
/// defines them structurally rather than textually.
/// </para>
/// </remarks>
abstract class Ast
{
	/// <summary>
	/// The offset in the pattern at which this node starts. Diagnostics only.
	/// </summary>
	public int PatternOffset { get; set; }

	/// <summary>
	/// Whether the tree holds a unified element anywhere.
	/// </summary>
	/// <remarks>
	/// This decides two things at once, which is why it lives here rather than with either of
	/// them. Without a unified element &#961; is the identity, so W3 holds for nothing and the
	/// canonical language is the accepted one.
	/// </remarks>
	/// <param name="node">The node to search from.</param>
	/// <returns>Whether one was found.</returns>
	public static bool ContainsUnified(Ast node)
	{
		switch (node)
		{
			case AstUnified:
				return true;

			case AstSequence sequence:
				foreach (Ast child in sequence.Children)
				{
					if (ContainsUnified(child)) { return true; }
				}

				return false;

			case AstAlternation alternation:
				foreach (Ast child in alternation.Children)
				{
					if (ContainsUnified(child)) { return true; }
				}

				return false;

			case AstOptional optional:
				return ContainsUnified(optional.Child);

			case AstInterval interval:
				return ContainsUnified(interval.Child);

			default:
				return false;
		}
	}
}

/// <summary>The empty string, written <c>()</c>.</summary>
sealed class AstEmpty : Ast
{
}

/// <summary>A set of characters matching one position, such as <c>A</c>, <c>\9</c> or <c>[A-F]</c>.</summary>
sealed class AstChars : Ast
{
	public AstChars(AsciiCharSet charSet)
	{
		this.CharSet = charSet;
	}

	public AsciiCharSet CharSet { get; }
}

/// <summary>A decimal range, written <c>#[</c><i>lo</i><c>-</c><i>hi</i><c>]</c>.</summary>
/// <remarks>
/// The digit counts are the counts <em>as written</em>, which is what fixes the widths
/// generated: <c>#[00-105]</c> does not match <c>7</c> while <c>#[0-105]</c> does.
/// </remarks>
sealed class AstDecimalRange : Ast
{
	public AstDecimalRange(ulong low, int lowDigitCount, ulong high, int highDigitCount)
	{
		this.Low = low;
		this.LowDigitCount = lowDigitCount;
		this.High = high;
		this.HighDigitCount = highDigitCount;
	}

	public ulong Low { get; }
	public int LowDigitCount { get; }
	public ulong High { get; }
	public int HighDigitCount { get; }
}

/// <summary>Two or more elements in sequence.</summary>
sealed class AstSequence : Ast
{
	public AstSequence(IReadOnlyList<Ast> children)
	{
		this.Children = children;
	}

	public IReadOnlyList<Ast> Children { get; }
}

/// <summary>Two or more alternatives separated by <c>|</c>.</summary>
sealed class AstAlternation : Ast
{
	public AstAlternation(IReadOnlyList<Ast> children)
	{
		this.Children = children;
	}

	public IReadOnlyList<Ast> Children { get; }
}

/// <summary>An optional element, written <c>x?</c>.</summary>
sealed class AstOptional : Ast
{
	public AstOptional(Ast child)
	{
		this.Child = child;
	}

	public Ast Child { get; }
}

/// <summary>A bounded interval, written <c>x{n}</c> or <c>x{m,n}</c>.</summary>
sealed class AstInterval : Ast
{
	public AstInterval(Ast child, int minCount, int maxCount)
	{
		this.Child = child;
		this.MinCount = minCount;
		this.MaxCount = maxCount;
	}

	public Ast Child { get; }
	public int MinCount { get; }
	public int MaxCount { get; }
}

/// <summary>
/// How a unified element was written, which is needed only so that a well-formedness
/// message can name the form the author used rather than the form it expands to.
/// </summary>
enum UnifiedForm
{
	/// <summary><c>x!y</c>.</summary>
	Explicit,
	/// <summary><c>x!!</c>, which expands to <c>x?!(x)</c>.</summary>
	Reproduced,
	/// <summary><c>x!?</c>, which expands to <c>x?!()</c>.</summary>
	Dropped,
	/// <summary>
	/// One branch of a case fold, such as the <c>[Aa]!A</c> that <c>\CA</c> expands to.
	/// </summary>
	/// <remarks>
	/// The form records where the branch came from and changes no rule. An outer fold reaches a
	/// fold branch like any other unified element, so where two folds meet the outer governs and
	/// <c>\C(AB\cC)</c> prints <c>ABC</c>.
	/// </remarks>
	Fold,
}

/// <summary>
/// A unified element, written <c>x!y</c>. Which of the strings the subject accepts was
/// matched is not part of the encoding, and the rendering is printed in its place.
/// </summary>
/// <remarks>
/// For the two abbreviated forms the subject is the <see cref="AstOptional"/> wrapping what
/// was written, so <see cref="Subject"/> is always the expression whose choice goes unencoded.
/// The <c>x!!</c> form shares one subtree between <see cref="Subject"/> and
/// <see cref="Rendering"/>; nothing in the tree is mutated after parsing, so that is safe.
/// </remarks>
sealed class AstUnified : Ast
{
	public AstUnified(Ast subject, Ast rendering, UnifiedForm form)
	{
		this.Subject = subject;
		this.Rendering = rendering;
		this.Form = form;
	}

	public Ast Subject { get; }
	public Ast Rendering { get; }
	public UnifiedForm Form { get; }
}

File: src/js/lib/ast.js
// Copyright (c) Tim Gordon.
// This file is licensed to you under the Apache Licence, Version 2.0. See the LICENSE file.

/**
 * A node in the abstract syntax tree of a naxp.
 *
 * The tree keeps the structure the source was written in. Intervals and digits ranges are
 * deliberately *not* expanded here. The cap on an interval count exists so that an implementation
 * can reject a naxp before expanding it, and expanding at parse time would throw that away:
 * `(A{99}){99}` is eleven characters of source and nearly ten thousand characters of expansion.
 *
 * Groups do not survive parsing. `(A)` and `A` give the same tree, and the two abbreviated
 * replaceable forms are expanded into the general one, since version 0.4 defines them structurally
 * rather than textually.
 */
export class Ast {
	constructor() {
		/**
		 * The offset in the source at which this node starts. Diagnostics only.
		 *
		 * @type {number}
		 */
		this.sourceOffset = 0;
	}
}

/** The empty string, written `()`. */
export class AstEmpty extends Ast {
}

/** A set of characters matching one position, such as `A`, `\9` or `[A-F]`. */
export class AstChars extends Ast {
	/**
	 * @param {import('./ascii-char-set.js').AsciiCharSet} charSet The characters.
	 */
	constructor(charSet) {
		super();
		this.charSet = charSet;
	}
}

/**
 * A digits range, written `#[`*lo*`-`*hi*`]`.
 *
 * The digit counts are the counts *as written*, which is what fixes the widths generated:
 * `#[00-105]` does not match `7` while `#[0-105]` does.
 *
 * The bounds are ordinary numbers rather than BigInts. A bound may have at most fifteen digits,
 * which the specification chose precisely because that is what a double holds exactly, so no bound
 * can reach the point where a number stops being able to represent it.
 */
export class AstDigitsRange extends Ast {
	/**
	 * @param {number} low The lower bound.
	 * @param {number} lowDigitCount How many digits the lower bound was written with.
	 * @param {number} high The upper bound.
	 * @param {number} highDigitCount How many digits the upper bound was written with.
	 */
	constructor(low, lowDigitCount, high, highDigitCount) {
		super();
		this.low = low;
		this.lowDigitCount = lowDigitCount;
		this.high = high;
		this.highDigitCount = highDigitCount;
	}
}

/** Two or more elements in sequence. */
export class AstSequence extends Ast {
	/**
	 * @param {Ast[]} children The elements, in order.
	 */
	constructor(children) {
		super();
		this.children = children;
	}
}

/** Two or more alternatives separated by `|`. */
export class AstAlternation extends Ast {
	/**
	 * @param {Ast[]} children The alternatives, in the order written.
	 */
	constructor(children) {
		super();
		this.children = children;
	}
}

/** An optional element, written `x?`. */
export class AstOptional extends Ast {
	/**
	 * @param {Ast} child The element that may be absent.
	 */
	constructor(child) {
		super();
		this.child = child;
	}
}

/** A bounded interval, written `x{n}` or `x{m,n}`. */
export class AstInterval extends Ast {
	/**
	 * @param {Ast} child The element repeated.
	 * @param {number} minCount The fewest repetitions.
	 * @param {number} maxCount The most repetitions.
	 */
	constructor(child, minCount, maxCount) {
		super();
		this.child = child;
		this.minCount = minCount;
		this.maxCount = maxCount;
	}
}

/**
 * How a replaceable element was written, which is needed only so that a well-formedness message
 * can name the form the author used rather than the form it expands to.
 *
 * @enum {string}
 */
export const ReplaceableForm = Object.freeze({
	/** `x!y`. */
	Explicit: 'Explicit',
	/** `x!!`, which expands to `x?!(x)`. */
	Reproduced: 'Reproduced',
	/** `x!?`, which expands to `x?!()`. */
	Dropped: 'Dropped',
});

/**
 * A replaceable element, written `x!y`. Which of the strings the subject accepts was matched is not
 * part of the encoding, and the rendering is printed in its place.
 *
 * For the two abbreviated forms the subject is the {@link AstOptional} wrapping what was written,
 * so `subject` is always the expression whose choice goes unencoded. The `x!!` form shares one
 * subtree between `subject` and `rendering`; nothing in the tree is mutated after parsing, so that
 * is safe.
 */
export class AstReplaceable extends Ast {
	/**
	 * @param {Ast} subject The expression whose choice goes unencoded.
	 * @param {Ast} rendering What is printed in its place.
	 * @param {string} form How it was written, one of {@link ReplaceableForm}.
	 */
	constructor(subject, rendering, form) {
		super();
		this.subject = subject;
		this.rendering = rendering;
		this.form = form;
	}
}

/**
 * Whether a tree holds a replaceable element anywhere.
 *
 * This decides two things at once, which is why it is one function rather than living with either
 * of them. Without a replaceable element ρ is the identity, so W3 holds for nothing and the
 * canonical language is the accepted one.
 *
 * @param {Ast} node The node to search from.
 * @returns {boolean} Whether one was found.
 */
export function containsReplaceable(node) {
	if (node instanceof AstReplaceable) { return true; }

	if (node instanceof AstSequence || node instanceof AstAlternation) {
		return node.children.some(containsReplaceable);
	}

	if (node instanceof AstOptional || node instanceof AstInterval) {
		return containsReplaceable(node.child);
	}

	return false;
}
